text/0019-custom-shaders.md

```diff
@@ -0,0 +1,299 @@
+                device,
+                staging_belt,
+                encoder,
+                &layer.custom_shader_quads,
+                &serializable_instances,
+                transformation,
+                scale_factor,
+                bounds,
+                target,
+            );
+        }
+```
+
+In `graphics::renderer`, we add a new method to the `Renderer` struct:
+```rs
+    fn make_custom_shader_quad(
+        &mut self,
+        quad: renderer::CustomShaderQuad,
+        background: impl Into<Background>,
+    ) {
+        self.primitives.push(Primitive::CustomShaderQuad {
+            bounds: quad.bounds,
+            background: background.into(),
+            mouse_position: Vector::new(
+                quad.mouse_position.x,
+                quad.mouse_position.y,
+            ),
+            mouse_click: Vector::new(quad.mouse_click.x, quad.mouse_click.y),
+            time: quad.time,
+            frame: quad.frame_number,
+            shader_code: quad.shader_code,
+        });
+    }
+```
+
+This `make_custom_shader_quad(..)` method must be called by the end user to instanciate a custom shader quad. It is also present inside the `draw(..)` method in the example above. In the case where the `mouse_position`, `mouse_click`, `time` and `frame` fields are not needed, the user must give dummy values to these fields anyway.
+
+The custom shader render pipeline is defined in `wgpu::custom_shader_quad`. This is where the shader code is inserted.
+
+For all widgets that implement the `Overlay` trait, one could use a custom shader by calling `make_custom_shader_quad(..)` instead of `fill_quad(..)` on the `Renderer` inside the `draw(..)` method of the `Overlay` trait.
+<br/><br/> 
+
+## Drawbacks
+
+Since the interaction information (mouse press, key press, etc) is sent through attributes, this information is sent once for every vertex, which amounts to four times the quantity of data that is actually needed. Ideally, this information would be sent through a uniform buffer that is unique to each shader quad, but this would require a more complex implementation and would most likely break the current implementation of other rendering primitives. In the grand scheme of things, this quantity of extra data  is not very large, but it is something to keep in mind since it wouldn't be practical to generalize the above scheme to more complex vertex geometries.
+
+Custom shaders would only be available with the `wgpu` feature. 
+
+The byte size of the data going from the CPU to the GPU is fixed through attributes is fixed at 26 bytes right now. Users cannot send more data than this to their custom shader.
+<br/><br/> 
+## Rationale and alternatives
+
+Currently, the alternative to custom shader quads is to implement a custom WGPU pipeline with a lot of boilerplate, but with more flexibility. The custom shader quads allow users who are not familiar with GPU pipelines to use custom shaders rather easily.
```

Comment:
I feel the more flexible solution may ultimately be necessary. It's pretty standard for UI toolkits (and browsers) to offer a way to use custom OpenGL code in a UI element.

Perhaps if this provided an `iced::widget::wgsl_shaded_quad` widget, that could offer a simple API for this use case whether that wraps a `CustomShaderQuad` primitive or provides the necessary boilerplate over a more flexible solution. Though if it makes sense to offer both as primitives that could work too.

Reply:
I've added a `native::widget::wgsl_shader_quads`, with the following fields
```rs
pub struct WgslShaderQuad<Message> {
    on_press: Option<Message>,
    on_hover_entering: Option<Message>,
    on_hover_leaving: Option<Message>,
    width: f32,
    height: f32,
    padding: Padding,
    time: Duration,
    handle: shader::Handle,
}
```
where `time` is sent to the GPU for animating the shader. The `draw(..)` method for the new widget calls `renderer.make_custom_shader_quad(..)`.

There's also a new example called `shader_widget`, which is now significantly shorter than the previous example since there's no need to define a custom widget anymore. 

The new widget does not add more flexibility in its current state. In fact, it takes flexibility away since the `WgslShaderQuad` widget is now fixed. So I'd like to discuss the specific changes that would be required to attain the level of flexibility that is expected here. For example, these are potential functionalities: 

Give user control over
1. uniform variables
2. attributes
3. instancing
4. textures
5. compute shaders